File: Mazfall_Zombie_Versions_04182026/Mazfall_Zombie_Alpha_v0.5.py
#!/usr/bin/env python3
# MAZFALL_ZOMBIE_ALPHA_v0.5 · Doomfall-0.1 soul × Mazfall-α0.02 spine
# ∞ recursive chunks · ∞ zombie respawn · ∞ ammo · chaingun overheat
# WASD/arrows + mouse · SPACE/LMB fire · N new seed · R respawn · ESC quit
import sys,subprocess,math,random,time
def _i(p):
    try:__import__(p)
    except:subprocess.call([sys.executable,"-m","pip","install",p,"--user","-q"])
for p in("pygame","numpy"):_i(p)
import pygame,numpy as np
W,H,S=100,34,10;FV=math.pi/3;FH=FV*.5;P2=math.pi*2;MD=40.;RM=" .:-=+*#%@";RL=9
CZ=21;CM=10;SR=22050;MV=.18;TR=.08;MS=.005  # faster move/turn/mouse
# ── LCG doom P_Random soul ─────────────────────────────────────────
def _cs(w,x,y):return((w^(x*1664525+1013904223)^(y*22695477+1))*2654435761)&0xFFFFFFFF
def _gc(w,cx,cy):
    v=_cs(w,cx,cy);M=[[1]*CZ for _ in range(CZ)]
    M[0][CM]=M[CZ-1][CM]=M[CM][0]=M[CM][CZ-1]=0
    def r():
        nonlocal v;v=(1664525*v+1013904223)&0xFFFFFFFF;return v
    D=[(0,-2),(0,2),(-2,0),(2,0)]
    def sd():
        d=list(D)
        for i in range(3,0,-1):j=r()%(i+1);d[i],d[j]=d[j],d[i]
        return d
    M[1][1]=0;st=[(1,1,sd())]
    while st:
        x,y,ds=st[-1]
        if not ds:st.pop();continue
        dx,dy=ds.pop(0);nx,ny=x+dx,y+dy
        if 1<=nx<CZ-1 and 1<=ny<CZ-1 and M[ny][nx]:
            M[y+dy//2][x+dx//2]=0;M[ny][nx]=0;st.append((nx,ny,sd()))
    return M
CH={}
def cell(ws,wx,wy):
    cx,lx=divmod(int(wx),CZ);cy,ly=divmod(int(wy),CZ);k=(cx,cy)
    if k not in CH:
        if len(CH)>256:CH.pop(next(iter(CH)))
        CH[k]=_gc(ws,cx,cy)
    return CH[k][ly][lx]
def cast(ws,px,py,a):
    dx,dy=math.cos(a),math.sin(a);t=.04
    while t<MD:
        if cell(ws,px+dx*t,py+dy*t):return t
        t+=.04
    return MD
# ── audio ─────────────────────────────────────────────────────────
pygame.mixer.pre_init(SR,-16,1,256);pygame.init()
def _snd(w,A=18000):return pygame.sndarray.make_sound((np.clip(w,-1,1)*A).astype(np.int16))
def _gun():
    d=.08;n=int(SR*d);t=np.linspace(0,d,n,False)
    return _snd(np.random.uniform(-1,1,n)*np.exp(-t*30)*.6+np.sin(P2*np.cumsum(80*np.exp(-t*6)+40)/SR)*np.exp(-t*8))
def _hit():
    d=.22;n=int(SR*d);t=np.linspace(0,d,n,False)
    return _snd(np.random.uniform(-1,1,n)*np.exp(-t*10)*.8,22000)
def _hurt():
    d=.18;n=int(SR*d);t=np.linspace(0,d,n,False)
    return _snd(np.sin(P2*140*t)*np.exp(-t*7)*.7+np.random.uniform(-1,1,n)*np.exp(-t*20)*.3)
SG=_gun();SX=_hit();SH_=_hurt()
scr=pygame.display.set_mode((W*S,H*S));pygame.display.set_caption("MAZFALL_ZOMBIE_ALPHA_v0.5")
fn=pygame.font.SysFont("Courier",S+2,bold=True);clk=pygame.time.Clock()
GC={}
def G(c,v):
    k=(c,v>>3)
    if k not in GC:GC[k]=fn.render(c,True,(v,v,v))
    return GC[k]
CC=[math.cos(-FH+x/W*FV)for x in range(W)];CS_=[math.sin(-FH+x/W*FV)for x in range(W)]
FC=[' 'if abs(y-H/2)/(H/2)<.1 else'.'if(y-H/2)/(H/2)<.45 else':'for y in range(H)]
ws=random.randint(0,0xFFFFFFFF);px,py,ang=1.5,1.5,0.
hp=100;sc=kl=0;lv=1;heat=0.;OH=False
MAXH=100.;HPS=8.;COOL=72.;RES=16.;RATE=.035;ls=lh=0.
def _far(px,py):
    for r in range(4,30,3):
        for _ in range(24):
            a=random.uniform(0,P2);d=r+random.random()*3
            wx,wy=px+math.cos(a)*d,py+math.sin(a)*d
            if not cell(ws,wx,wy):return wx,wy
    return px+5,py+5
Z=[]
def _spawn():
    wx,wy=_far(px,py);Z.append([wx,wy,1+lv//3])
for _ in range(40):_spawn()
pygame.event.set_grab(True);pygame.mouse.set_visible(False)
run=True
while run:
    dt=clk.tick(30)/1000.;now=time.time()
    for e in pygame.event.get():
        if e.type==pygame.QUIT:run=False
        elif e.type==pygame.KEYDOWN:
            if e.key==pygame.K_ESCAPE:run=False
            elif e.key==pygame.K_n:
                ws=random.randint(0,0xFFFFFFFF);CH.clear();px,py=1.5,1.5;Z.clear()
                for _ in range(40+lv*4):_spawn()
            elif e.key==pygame.K_r and hp<=0:
                hp=100;Z.clear()
                for _ in range(40+lv*4):_spawn()
        elif e.type==pygame.MOUSEMOTION and hp>0:ang+=e.rel[0]*MS
    ks=pygame.key.get_pressed()
    if hp>0:
        ca,sa=math.cos(ang),math.sin(ang);fx,fy=ca*MV,sa*MV;rx_,ry_=-sa*MV,ca*MV;dx=dy=0.
        if ks[pygame.K_w]or ks[pygame.K_UP]:dx+=fx;dy+=fy
        if ks[pygame.K_s]or ks[pygame.K_DOWN]:dx-=fx;dy-=fy
        if ks[pygame.K_a]:dx-=rx_;dy-=ry_
        if ks[pygame.K_d]:dx+=rx_;dy+=ry_
        if ks[pygame.K_LEFT]:ang-=TR
        if ks[pygame.K_RIGHT]:ang+=TR
        if not cell(ws,px+dx,py):px+=dx
        if not cell(ws,px,py+dy):py+=dy
        # ── CHAINGUN · ∞ ammo · heat spine ──────────────────────
        if(ks[pygame.K_SPACE]or pygame.mouse.get_pressed()[0])and not OH and now-ls>RATE:
            ls=now;SG.play();heat=min(MAXH,heat+HPS)
            if heat>=MAXH:OH=True
            wd=cast(ws,px,py,ang);best=-1;bd=1e9
            for i,z in enumerate(Z):
                ex,ey=z[0]-px,z[1]-py;d2=ex*ex+ey*ey
                if d2>wd*wd:continue
                d=d2**.5;a=(math.atan2(ey,ex)-ang+math.pi)%P2-math.pi
                if abs(a)<.10+.40/(d+.5)and d<bd:best=i;bd=d
            if best>=0:
                Z[best][2]-=1
                if Z[best][2]<=0:
                    Z.pop(best);SX.play();sc+=10;kl+=1;_spawn();_spawn()  # ∞ recursive: 2-for-1
        if heat>0:heat=max(0.,heat-COOL*dt)
        if OH and heat<RES:OH=False
        # ── zombie AI — tier grows with lv ──────────────────────
        ZSP=.035+lv*.006
        for z in Z:
            ex,ey=px-z[0],py-z[1];d2=ex*ex+ey*ey
            if d2<.0001:continue
            d=d2**.5;vx,vy=ex/d*ZSP,ey/d*ZSP
            if not cell(ws,z[0]+vx,z[1]):z[0]+=vx
            if not cell(ws,z[0],z[1]+vy):z[1]+=vy
            if d2<.49 and now-lh>.8:hp-=2+lv//2;lh=now;SH_.play()
        if kl>=lv*20:lv+=1;[_spawn()for _ in range(3)]
        if hp<=0:SH_.play()
    # ── render · Doomfall-0.1 minimalist soul ─────────────────────
    scr.fill((0,0,0));ca_=math.cos(ang);sa_=math.sin(ang);zb=[MD]*W
    for x in range(W):
        co=CC[x];so=CS_[x];ra=math.atan2(sa_*co+ca_*so,ca_*co-sa_*so)
        d=cast(ws,px,py,ra)*co;zb[x]=d
        wh=int(H/(d+.1));top=(H-wh)>>1
        ch=RM[max(0,min(RL,int((1-d/MD)*RL)))];v=max(60,int(220*(1-d/MD)))
        for y in range(H):
            if top<=y<top+wh:scr.blit(G(ch,v),(x*S,y*S))
            elif y>top+wh and FC[y]!=' ':scr.blit(G(FC[y],95),(x*S,y*S))
    # sprites · painter-sorted · zbuf-gated
    for z in sorted(Z,key=lambda z:-((z[0]-px)**2+(z[1]-py)**2)):
        rx,ry=z[0]-px,z[1]-py;d=math.hypot(rx,ry)
        if d<.2:continue
        a=(math.atan2(ry,rx)-ang+math.pi)%P2-math.pi
        if abs(a)>FH+.15:continue
        col=int((a/FV+.5)*W);sh=int(H/(d+.1));sw=max(1,int(sh*.55))
        top=max(0,(H-sh)>>1);lft=col-(sw>>1)
        for xx in range(max(0,lft),min(W,lft+sw)):
            if d>=zb[xx]:continue
            rxr=(xx-lft)/max(1,sw)
            for yy in range(top,min(H,top+sh)):
                ryr=(yy-top)/max(1,sh)
                if ryr<.2 and.35<rxr<.65:g='o'
                elif ryr<.72 and.18<rxr<.82:g='z'
                elif ryr>=.72 and rxr<.4:g='/'
                elif ryr>=.72 and rxr>.6:g='\\'
                elif ryr>=.72:g='|'
                else:g=None
                if g:scr.blit(G(g,220),(xx*S,yy*S))
    if hp>0:scr.blit(G('+',220),(W*S//2,H*S//2))
    # heat bar · 8-slot eightfold soul
    bw=int(W*S*.22);bh=8;bx=(W*S-bw)//2;by=H*S//2+S*2
    pygame.draw.rect(scr,(40,40,40),(bx,by,bw,bh))
    hf=int(bw*heat/MAXH)
    if hf>0:
        c=(255,80,80)if OH else(255,200,60)if heat>60 else(80,220,80)
        if OH and int(now*8)%2:c=(255,200,50)
        pygame.draw.rect(scr,c,(bx,by,hf,bh))
    hud=f"HP:{hp:3d} SC:{sc} K:{kl} LV:{lv} Z:{len(Z)} {'[OVERHEAT]'if OH else f'[{int(heat):02d}%]'}"
    scr.blit(fn.render(hud,True,(120,255,120)if hp>50 else(255,220,80)if hp>20 else(255,80,80)),(8,4))
    if hp<=0:
        m=fn.render("ELIMINATED — R:RESPAWN  N:NEW SEED",True,(255,60,60))
        scr.blit(m,(W*S//2-m.get_width()//2,H*S//2+60))
    pygame.display.flip()
pygame.quit();sys.exit()
# .- ∞ recurse · the maze has no exit

File: Mazfall_Zombie_Versions_04182026/Mazfall_Zombie_Alpha_v0.5.c
/* MAZFALL_ZOMBIE_ALPHA_v0.5  —  pure C · ANSI terminal · POSIX
 * Doomfall-0.1 ASCII ramp × Mazfall-α0.02 infinite chunked LCG world
 * WASD move · JL turn · SPACE fire · N new seed · R respawn · Q quit
 * build:  cc -O2 -lm Mazfall_Zombie_Alpha_v0.5.c -o mz05
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <sys/select.h>
#include <sys/ioctl.h>

#define W 100
#define H 34
#define CZ 21
#define CM 10
#define MD 40.0
#define CACHE 256
#define ZMAX 256
#define P2 6.28318530717958647692
#define FV 1.0471975511965976
#define FH 0.5235987755982988
static const char RM[] = " .:-=+*#%@";
#define RL 9

/* ── LCG — doom P_Random soul ─────────────────────────────────── */
typedef struct { unsigned int v; } lcg_t;
static unsigned int cseed(unsigned int w, int cx, int cy) {
    unsigned int v = w ^ ((unsigned)cx*1664525u+1013904223u) ^ ((unsigned)cy*22695477u+1u);
    return v * 2654435761u;
}
static unsigned int lrn(lcg_t *l) { l->v = l->v*1664525u + 1013904223u; return l->v; }

/* ── chunk gen — recursive backtracker maze ───────────────────── */
typedef unsigned char cell_t;
typedef struct { int cx, cy; cell_t m[CZ][CZ]; int hit; } chunk_t;
static chunk_t CH[CACHE];
static int CH_n = 0;
static int CH_tick = 0;

static void gen_chunk(unsigned int ws, int cx, int cy, chunk_t *dst) {
    lcg_t l; l.v = cseed(ws, cx, cy);
    dst->cx = cx; dst->cy = cy;
    for (int y = 0; y < CZ; y++) for (int x = 0; x < CZ; x++) dst->m[y][x] = 1;
    dst->m[0][CM] = dst->m[CZ-1][CM] = dst->m[CM][0] = dst->m[CM][CZ-1] = 0;
    /* iterative DFS carver — preserve doom carve_iter soul */
    int stack_x[CZ*CZ], stack_y[CZ*CZ], sp = 0;
    dst->m[1][1] = 0; stack_x[sp]=1; stack_y[sp]=1; sp++;
    int dd[4][2] = {{0,-2},{0,2},{-2,0},{2,0}};
    while (sp > 0) {
        int x = stack_x[sp-1], y = stack_y[sp-1];
        int order[4] = {0,1,2,3};
        for (int i = 3; i > 0; i--) { int j = lrn(&l) % (i+1); int t=order[i]; order[i]=order[j]; order[j]=t; }
        int moved = 0;
        for (int k = 0; k < 4; k++) {
            int dx = dd[order[k]][0], dy = dd[order[k]][1];
            int nx = x+dx, ny = y+dy;
            if (nx>=1 && nx<CZ-1 && ny>=1 && ny<CZ-1 && dst->m[ny][nx]) {
                dst->m[y+dy/2][x+dx/2] = 0; dst->m[ny][nx] = 0;
                stack_x[sp]=nx; stack_y[sp]=ny; sp++; moved = 1; break;
            }
        }
        if (!moved) sp--;
    }
}

static chunk_t *get_chunk(unsigned int ws, int cx, int cy) {
    for (int i = 0; i < CH_n; i++) if (CH[i].cx == cx && CH[i].cy == cy) { CH[i].hit = ++CH_tick; return &CH[i]; }
    int slot;
    if (CH_n < CACHE) slot = CH_n++;
    else { int lo = CH[0].hit, li = 0; for (int i=1;i<CH_n;i++) if (CH[i].hit<lo){lo=CH[i].hit;li=i;} slot = li; }
    gen_chunk(ws, cx, cy, &CH[slot]);
    CH[slot].hit = ++CH_tick;
    return &CH[slot];
}

static inline int imod(int a, int n) { int r = a % n; return r < 0 ? r + n : r; }
static inline int idiv(int a, int n) { int r = a / n; if ((a % n) < 0) r--; return r; }

static int cell(unsigned int ws, double wx, double wy) {
    int ix = (int)floor(wx), iy = (int)floor(wy);
    int cx = idiv(ix, CZ), cy = idiv(iy, CZ);
    int lx = imod(ix, CZ), ly = imod(iy, CZ);
    return get_chunk(ws, cx, cy)->m[ly][lx];
}

/* ── DDA raycast ──────────────────────────────────────────────── */
static double cast(unsigned int ws, double px, double py, double a) {
    double dx = cos(a), dy = sin(a), t = 0.04;
    while (t < MD) { if (cell(ws, px+dx*t, py+dy*t)) return t; t += 0.04; }
    return MD;
}

/* ── terminal I/O — raw mode ─────────────────────────────────── */
static struct termios orig_tio;
static void tio_restore(void) {
    tcsetattr(STDIN_FILENO, TCSANOW, &orig_tio);
    printf("\x1b[?25h\x1b[0m\x1b[2J\x1b[H"); fflush(stdout);
}
static void tio_raw(void) {
    tcgetattr(STDIN_FILENO, &orig_tio);
    struct termios t = orig_tio;
    t.c_lflag &= ~(ICANON | ECHO);
    t.c_cc[VMIN] = 0; t.c_cc[VTIME] = 0;
    tcsetattr(STDIN_FILENO, TCSANOW, &t);
    printf("\x1b[?25l\x1b[2J"); fflush(stdout);
    atexit(tio_restore);
}
static int kbhit_read(char *c) { return read(STDIN_FILENO, c, 1) == 1; }

/* ── zombie pool ──────────────────────────────────────────────── */
typedef struct { double x, y; int hp, alive; } zomb_t;
static zomb_t Z[ZMAX];
static int ZN = 0;

static void far_floor(unsigned int ws, double px, double py, double *ox, double *oy) {
    for (int r = 4; r < 30; r += 3) for (int k = 0; k < 24; k++) {
        double a = (rand()/(double)RAND_MAX)*P2;
        double d = r + (rand()/(double)RAND_MAX)*3.0;
        double wx = px + cos(a)*d, wy = py + sin(a)*d;
        if (!cell(ws, wx, wy)) { *ox = wx; *oy = wy; return; }
    }
    *ox = px+5; *oy = py+5;
}

static void spawn_zomb(unsigned int ws, double px, double py, int lv) {
    if (ZN >= ZMAX) return;
    far_floor(ws, px, py, &Z[ZN].x, &Z[ZN].y);
    Z[ZN].hp = 1 + lv/3; Z[ZN].alive = 1; ZN++;
}

/* ── render buffer ────────────────────────────────────────────── */
static char BUF[H][W+1];
static unsigned char SHD[H][W];
static double ZB[W];
static double CC_[W], CS_[W];

/* ── nano-sleep for ~20fps cap ───────────────────────────────── */
static double now_s(void) { struct timespec ts; clock_gettime(CLOCK_MONOTONIC,&ts); return ts.tv_sec+ts.tv_nsec*1e-9; }

int main(void) {
    srand((unsigned)time(NULL));
    unsigned int ws = (unsigned int)(((unsigned long)rand()<<16) ^ rand() ^ time(NULL));
    double px = 1.5, py = 1.5, ang = 0.0;
    int hp = 100, sc = 0, kl = 0, lv = 1;
    double heat = 0.0; int OH = 0;
    const double MAXH=100, HPS=8, COOL=72, RES=16, RATE=0.035;
    const double MV=0.22, TR=0.11;
    double ls = 0, lh = 0;
    for (int x = 0; x < W; x++) { CC_[x]=cos(-FH+(double)x/W*FV); CS_[x]=sin(-FH+(double)x/W*FV); }
    for (int i = 0; i < 40; i++) spawn_zomb(ws, px, py, lv);
    tio_raw();
    double last = now_s();
    int run = 1;
    while (run) {
        double t = now_s(); double dt = t - last; last = t;
        /* input — drain kbd buffer */
        char c;
        while (kbhit_read(&c)) {
            if (c == 27 || c == 'q' || c == 'Q') run = 0;
            else if (c == 'w' || c == 'W') {
                double nx = px + cos(ang)*MV, ny = py + sin(ang)*MV;
                if (!cell(ws,nx,py)) px = nx;
                if (!cell(ws,px,ny)) py = ny;
            } else if (c == 's' || c == 'S') {
                double nx = px - cos(ang)*MV, ny = py - sin(ang)*MV;
                if (!cell(ws,nx,py)) px = nx;
                if (!cell(ws,px,ny)) py = ny;
            } else if (c == 'a' || c == 'A') {
                double nx = px - (-sin(ang))*MV, ny = py - cos(ang)*MV;
                if (!cell(ws,nx,py)) px = nx;
                if (!cell(ws,px,ny)) py = ny;
            } else if (c == 'd' || c == 'D') {
                double nx = px + (-sin(ang))*MV, ny = py + cos(ang)*MV;
                if (!cell(ws,nx,py)) px = nx;
                if (!cell(ws,px,ny)) py = ny;
            } else if (c == 'j' || c == 'J') ang -= TR;
            else if (c == 'l' || c == 'L') ang += TR;
            else if (c == ' ') {
                if (!OH && hp > 0 && t - ls > RATE) {
                    ls = t;
                    heat += HPS; if (heat >= MAXH) { heat = MAXH; OH = 1; }
                    /* find closest target in cone */
                    double wd = cast(ws, px, py, ang);
                    int best = -1; double bd = 1e9;
                    for (int i = 0; i < ZN; i++) {
                        if (!Z[i].alive) continue;
                        double ex = Z[i].x-px, ey = Z[i].y-py, d2 = ex*ex+ey*ey;
                        if (d2 > wd*wd) continue;
                        double d = sqrt(d2);
                        double a = atan2(ey,ex) - ang; while(a>M_PI)a-=P2; while(a<-M_PI)a+=P2;
                        if (fabs(a) < 0.10 + 0.40/(d+0.5) && d < bd) { best = i; bd = d; }
                    }
                    if (best >= 0) {
                        Z[best].hp--;
                        if (Z[best].hp <= 0) {
                            Z[best].alive = 0; sc += 10; kl++;
                            /* compact array */
                            for (int i = best; i < ZN-1; i++) Z[i] = Z[i+1];
                            ZN--;
                            spawn_zomb(ws, px, py, lv); spawn_zomb(ws, px, py, lv); /* ∞ 2-for-1 */
                        }
                    }
                }
            } else if (c == 'n' || c == 'N') {
                ws = (unsigned int)((rand()<<16) ^ rand() ^ time(NULL));
                CH_n = 0; px = 1.5; py = 1.5; ZN = 0;
                for (int i = 0; i < 40+lv*4; i++) spawn_zomb(ws, px, py, lv);
            } else if ((c == 'r' || c == 'R') && hp <= 0) {
                hp = 100; ZN = 0;
                for (int i = 0; i < 40+lv*4; i++) spawn_zomb(ws, px, py, lv);
            }
        }
        /* heat decay */
        if (heat > 0) { heat -= COOL*dt; if (heat < 0) heat = 0; }
        if (OH && heat < RES) OH = 0;
        /* zombie AI */
        double zsp = 0.035 + lv*0.006;
        if (hp > 0) {
            for (int i = 0; i < ZN; i++) {
                double ex = px-Z[i].x, ey = py-Z[i].y, d2 = ex*ex+ey*ey;
                if (d2 < 1e-4) continue;
                double d = sqrt(d2), vx = ex/d*zsp, vy = ey/d*zsp;
                if (!cell(ws, Z[i].x+vx, Z[i].y)) Z[i].x += vx;
                if (!cell(ws, Z[i].x, Z[i].y+vy)) Z[i].y += vy;
                if (d2 < 0.49 && t - lh > 0.8) { hp -= 2 + lv/2; lh = t; }
            }
            if (kl >= lv*20) { lv++; for (int k = 0; k < 3; k++) spawn_zomb(ws, px, py, lv); }
        }
        /* ── render ─────────────────────────────────────────── */
        for (int y = 0; y < H; y++) { for (int x = 0; x < W; x++) { BUF[y][x] = ' '; SHD[y][x] = 0; } BUF[y][W] = 0; }
        double ca = cos(ang), sa = sin(ang);
        for (int x = 0; x < W; x++) {
            double co = CC_[x], so = CS_[x];
            double ra = atan2(sa*co+ca*so, ca*co-sa*so);
            double d = cast(ws, px, py, ra) * co;
            ZB[x] = d;
            int wh = (int)(H/(d+0.1)), top = (H-wh)>>1;
            int si = (int)((1.0 - d/MD) * RL); if (si<0)si=0; if (si>RL)si=RL;
            char ch = RM[si];
            int v = (int)(220*(1.0 - d/MD)); if (v<60) v=60;
            for (int y = 0; y < H; y++) {
                if (y >= top && y < top+wh) { BUF[y][x] = ch; SHD[y][x] = (unsigned char)v; }
                else if (y > top+wh) {
                    double fr = ((double)y-H/2.0)/(H/2.0);
                    char fc = fr<0.1?' ':fr<0.45?'.':':';
                    if (fc != ' ') { BUF[y][x] = fc; SHD[y][x] = 95; }
                }
            }
        }
        /* sprites — painter sort by insertion */
        int idx[ZMAX]; double dsq[ZMAX]; int N = 0;
        for (int i = 0; i < ZN; i++) {
            double rx = Z[i].x-px, ry = Z[i].y-py;
            dsq[N] = rx*rx+ry*ry; idx[N] = i; N++;
        }
        /* insertion sort descending by distance */
        for (int i = 1; i < N; i++) {
            double k = dsq[i]; int ki = idx[i]; int j = i-1;
            while (j >= 0 && dsq[j] < k) { dsq[j+1]=dsq[j]; idx[j+1]=idx[j]; j--; }
            dsq[j+1] = k; idx[j+1] = ki;
        }
        for (int n = 0; n < N; n++) {
            int i = idx[n];
            double rx = Z[i].x-px, ry = Z[i].y-py, d = sqrt(rx*rx+ry*ry);
            if (d < 0.2) continue;
            double a = atan2(ry,rx) - ang; while(a>M_PI)a-=P2; while(a<-M_PI)a+=P2;
            if (fabs(a) > FH+0.15) continue;
            int col = (int)((a/FV+0.5)*W);
            int sh = (int)(H/(d+0.1)); int sw = (int)(sh*0.55); if (sw<1)sw=1;
            int top = (H-sh)>>1; if (top<0) top=0;
            int lft = col - (sw>>1);
            for (int xx = lft>0?lft:0; xx < lft+sw && xx < W; xx++) {
                if (d >= ZB[xx]) continue;
                double rxr = (double)(xx-lft)/(sw>0?sw:1);
                for (int yy = top; yy < top+sh && yy < H; yy++) {
                    double ryr = (double)(yy-top)/(sh>0?sh:1);
                    char g = 0;
                    if (ryr<0.2 && rxr>0.35 && rxr<0.65) g='o';
                    else if (ryr<0.72 && rxr>0.18 && rxr<0.82) g='z';
                    else if (ryr>=0.72 && rxr<0.4) g='/';
                    else if (ryr>=0.72 && rxr>0.6) g='\\';
                    else if (ryr>=0.72) g='|';
                    if (g) { BUF[yy][xx] = g; SHD[yy][xx] = 220; }
                }
            }
        }
        /* crosshair */
        if (hp > 0) { BUF[H/2][W/2] = '+'; SHD[H/2][W/2] = 220; }
        /* flush — ANSI 256 gray ramp, home cursor */
        printf("\x1b[H");
        for (int y = 0; y < H; y++) {
            int last_v = -1;
            for (int x = 0; x < W; x++) {
                int v = SHD[y][x];
                if (v != last_v) {
                    if (v == 0) printf("\x1b[0m");
                    else { int g = 232 + (v*23)/255; if (g<232)g=232; if(g>255)g=255; printf("\x1b[38;5;%dm", g); }
                    last_v = v;
                }
                putchar(BUF[y][x] ? BUF[y][x] : ' ');
            }
            printf("\x1b[0m\n"); last_v = 0;
        }
        /* HUD */
        int bw = 22; int hf = (int)(bw*heat/MAXH);
        printf("\x1b[0mHP:%3d  SC:%d  K:%d  LV:%d  Z:%d  [", hp, sc, kl, lv, ZN);
        for (int i = 0; i < bw; i++) putchar(i<hf?'#':'.');
        printf("] %s  heat=%02d%%%s\n",
            OH ? "OVERHEAT" : "CHAINGUN",
            (int)heat,
            hp<=0 ? "   *** ELIMINATED — R:respawn  N:new seed  Q:quit ***" : "");
        fflush(stdout);
        /* ~20fps */
        double sleep = 0.05 - (now_s() - t);
        if (sleep > 0) { struct timespec ts; ts.tv_sec=0; ts.tv_nsec=(long)(sleep*1e9); nanosleep(&ts, NULL); }
    }
    return 0;
}
/* .- ∞ recurse · the maze has no exit */
